### PR TITLE
Move Allocator.h to c10

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14022 Move Scalar and ScalarType to c10/core&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13015236/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14023 Fix include paths for Scalar.h and ScalarType.h&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081609/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14024 Move ScalarTypeUtils.h to c10&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081604/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14058 Move UniqueVoidPtr to c10&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081602/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#14059 Move Allocator.h to c10**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081606/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14060 Fix include paths for Allocator.h&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081605/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14061 Move Storage and StorageImpl to c10&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081608/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14062 Fix include paths for Storage.h and StorageImpl.h&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081603/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14327 Move TensorTypeId to c10/core&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13131338/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14070 Fix include paths for TensorTypeId.h and TensorTypeIdRegistration.h&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13081610/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14141 Remove fake dependencies from TensorImpl to caffe2&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13113129/)

Differential Revision: [D13081606](https://our.internmc.facebook.com/intern/diff/D13081606/)

### CODE REVIEW

File: aten/src/ATen/core/Type.h [hunk 1]
```diff
@@ -29,10 +29,13 @@
 #endif
 #endif
 
+namespace c10 {
+struct Allocator;
```
Comment (ezyang) [inline aten/src/ATen/core/Type.h]: Given that this includes Allocator.h, I don't think there's any point forward declaring it...

File: c10/core/Allocator.h [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+  }
+};
+
+// Question: is this still needed?
+struct C10_API InefficientStdFunctionContext {
+  std::unique_ptr<void, std::function<void(void*)>> ptr_;
+  InefficientStdFunctionContext(
+      std::unique_ptr<void, std::function<void(void*)>>&& ptr)
+      : ptr_(std::move(ptr)) {}
+  static at::DataPtr makeDataPtr(
```
Comment (ezyang) [inline c10/core/Allocator.h]: at::

File: c10/core/Allocator.h [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+/** Set the allocator for DeviceType `t`. The passed in allocator pointer is
+ *  expected to have static lifetime; this function does NOT take ownership
+ *  of the raw pointer. (The reason for this is to prevent existing pointers
+ *  to an allocator of a particular device from being invalidated when
+ *  SetAllocator is called.)
+ *
+ *  Also note that this is not thraed-safe, and we assume this function will
+ *  only be called during initialization.
+ */
+C10_API void SetAllocator(at::DeviceType t, at::Allocator* alloc);
```
Comment (ezyang) [inline c10/core/Allocator.h]: at:: and a lot more

**Review by ezyang — APPROVED** [review]
accepted but with comments

**Comment by smessmer** [comment]
comments addressed in https://github.com/pytorch/pytorch/pull/14434

**Pull request closed without merging**